gpkit/constraints/set.py

```diff
@@ -0,0 +1,227 @@
+                if cstr is None:
+                    cstr = try_str_without(constraint, excluded)
+            else:
+                cstr = try_str_without(constraint, excluded)
+            if cstr[:8] != "        ":  # require indentation
+                cstr = "        " + cstr
+            lines.append(cstr)
+        return "\n".join(lines)
+
+    def latex(self, excluded=None):
+        "LaTeX representation of a ConstraintSet."
+        if not excluded:
+            excluded = ["units"]
+        if "cost" in excluded:
+            lines = []  # start with nothing
+        else:
+            excluded.append("cost")
+            lines = ["\\begin{array}[ll]", "\\text{}"]
+            if hasattr(self, "cost"):
+                lines.extend(["\\text{minimize}",
+                              "    & %s \\\\" % self.cost.latex(excluded),
+                              "\\text{subject to}"])
+        for constraint in self:
+            if hasattr(constraint, "_subconstr_latex"):
+                cstr = constraint._subconstr_latex(excluded)
+                if cstr is None:
+                    cstr = constraint.latex(excluded)
+            else:
+                cstr = constraint.latex(excluded)
+            if cstr[:6] != "    & ":  # require indentation
+                cstr = "    & " + cstr
+            lines.append(cstr)
+        lines.append("\\end{array}")
+        return "\n".join(lines)
+
+    def _subconstr_str(self, excluded=None):
+        "The collapsed appearance of a ConstraintSet"
+        pass
+
+    def _subconstr_tex(self, excluded=None):
+        "The collapsed appearance of a ConstraintSet"
+        pass
+
+    def flat(self, constraintsets=True):
+        "Yields contained constraints, optionally including constraintsets."
+        for constraint in self:
+            if not isinstance(constraint, ConstraintSet):
+                yield constraint
+            else:
+                if constraintsets:
+                    yield constraint
+                subgenerator = constraint.flat(constraintsets)
+                for yielded_constraint in subgenerator:
+                    yield yielded_constraint
+
+    def sub(self, subs, value=None):
+        "Substitutes in place."
+        if hasattr(self, "cost"):
+            self.cost = self.cost.sub(subs, value)
+        for i, constraint in enumerate(self):
+            self[i] = constraint.sub(subs, value)
+        return self
+
+    @property
+    def varkeys(self):
+        "Varkeys present in the constraints"
+        out = KeySet()
+        if hasattr(self, "cost"):
+            out.update(self.cost.varkeys)
+        for constraint in self:
+            if hasattr(constraint, "varkeys"):
+                out.update(constraint.varkeys)
+        return out
+
+    def as_posyslt1(self):
+        "Returns list of posynomials which must be kept <= 1"
+        posylist, self.posymap = [], []
+        for constraint in self:
+            constraint.substitutions = KeyDict()
```

Comment:
Right now it's not safe to use the same ConstraintSet in multiple models. (this isn't the only case where it's unsafe; fundamentally this is because `ConstraintSet.sub` modifies in place instead of returning a new `ConstraintSet`). This isn't new to the refactor.